### PR TITLE
chore: app for warehouse

### PR DESCRIPTION
**Pull request by achettyiitr**
# Description

- Introduce the warehouse app.
- Remove globals from the warehouse package.
- Added test cases.

## Linear Ticket

- https://linear.app/rudderstack/issue/PIPE-218/introduce-warehouse-app

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: warehouse/warehouse.go
Comment (achettyiitr) [inline warehouse/warehouse.go:316]: No point in throwing an error, since later we are already skipping the start for the warehouse service.

File: warehouse/trigger/trigger.go
Comment (fracasula) [inline warehouse/trigger/trigger.go:1]: One concern about this package is that it doesn't have anything to do with triggers, it's basically just a `sync.Map`, right?

So I was going to suggest to move it to some `util` package but if I benchmark it first.

```go
package trigger

import (
	"sync"
	"testing"
)

func BenchmarkTrigger(b *testing.B) {
	b.Run("trigger", func(b *testing.B) {
		store := NewStore()

		var wg sync.WaitGroup
		wg.Add(b.N * 3)

		b.ResetTimer()
		for i := 0; i < b.N; i++ {
			go func() {
				defer wg.Done()
				store.Trigger("concurrent")
			}()
			go func() {
				defer wg.Done()
				_ = store.IsTriggered("concurrent")
			}()
			go func() {
				defer wg.Done()
				store.ClearTrigger("concurrent")
			}()
		}

		wg.Wait()
	})
	b.Run("sync map", func(b *testing.B) {
		m := sync.Map{}

		var wg sync.WaitGroup
		wg.Add(b.N * 3)

		b.ResetTimer()
		for i := 0; i < b.N; i++ {
			go func() {
				defer wg.Done()
				m.Store("concurrent", struct{}{})
			}()
			go func() {
				defer wg.Done()
				m.Load("concurrent")
			}()
			go func() {
				defer wg.Done()
				m.Delete("concurrent")
			}()
		}

		wg.Wait()
	})
}
```

```
BenchmarkTrigger/trigger-20         	  720672	      1694 ns/op
BenchmarkTrigger/sync_map-20        	 2196574	      534.1 ns/op
```

File: warehouse/router.go
Comment (fracasula) [inline warehouse/router.go:737]: [very minor] it's simple enough in this case to avoid keeping the lock while calculating this:

```go
r.createJobMarkerMapLock.RLock()
lastCreatedAt, ok := r.createJobMarkerMap[warehouse.Identifier]
r.createJobMarkerMapLock.RUnlock()
```

File: warehouse/slave_worker_job.go
Comment (fracasula) [inline warehouse/slave_worker_job.go:393]:
```suggestion
	return batchID+"-"+tableName
```

File: warehouse/warehouse.go
Comment (fracasula) [inline warehouse/warehouse.go:287]: Same comment as above about the naming. If there is no `Stop` method please consider naming this `Run`.

File: warehouse/warehouse_test.go
Comment (fracasula) [inline warehouse/warehouse_test.go:175]: Can't we skip this step and pass `pgResource.Port` directly in the `c.Set()` below?

File: warehouse/trigger/trigger.go
Comment (achettyiitr) [reply warehouse/trigger/trigger.go:1]: Yeah, it has nothing to do with triggers, Just a plain vanilla map. Moving it to utils package.

Comment (achettyiitr) [reply warehouse/trigger/trigger.go:1]: Thanks for doing a quick Benchmark. 😄

File: warehouse/warehouse_test.go
Comment (fracasula) [inline warehouse/warehouse_test.go:251]: Why without env vars? It looks like we're passing them :thinking:

File: warehouse/warehouse_test.go
Comment (fracasula) [reply warehouse/warehouse_test.go:165]: Same with other things, for example if master then we start the cron archiver. Are we verifying that expectation somewhere?

**Review by fracasula — COMMENTED** [review]
Great work so far @achettyiitr, it's looking good :+1:

File: warehouse/warehouse_test.go
Comment (achettyiitr) [reply warehouse/warehouse_test.go:251]: I will update the test context. It was for `without warehouse env vars`. In that case, we are using `misc.GetConnectionString`

**Comment by achettyiitr** [comment]
> It still has nothing to do with triggers though so maybe we can give it a generic name or remove it completely.
> Thoughts?

Yeah, we can remove this and just pass the dependency for `triggerStore: syncs.Map`

File: warehouse/warehouse.go
Comment (achettyiitr) [reply warehouse/warehouse.go:287]: Thanks for pointing this out. 😄

File: warehouse/warehouse_test.go
Comment (achettyiitr) [reply warehouse/warehouse_test.go:175]: Since we are getting config.GetInt, I was expecting to set config.Set(<>, <int>). But later realized that [viper uses `cast.ToInt`](https://github.com/spf13/viper/blob/336e9f0a93eec4cec432eadda15d8d0803997d09/viper.go#L987) behind the scenes.

Comment (achettyiitr) [reply warehouse/warehouse_test.go:175]: Changing now.

Comment (achettyiitr) [reply warehouse/warehouse_test.go:175]: Thanks for pointing out.

File: warehouse/mode.go
Comment (achettyiitr) [reply warehouse/mode.go:5]: I am not sure about this. The rudder server also has a degraded mode which is different from the one kept in the warehouse. In the case of the warehouse, it is pointing to `Warehouse.runningMode`

File: warehouse/jobs/runner.go
Comment (achettyiitr) [reply warehouse/jobs/runner.go:119]: > Do we need the group in the first place? Since it is only really using a single routine, shouldn't we just remove it and block at the end with a return a.startAsyncJobRunner(ctx)?

> It would be great if we can have it accept a context as well. Then once the context is canceled Run can return after gracefully terminating all the routines that it started.

I will try to take these up in a separate PR. I have updated the [linear ticket](https://linear.app/rudderstack/issue/PIPE-168/jobs-cleanup) regarding this.

File: warehouse/warehouse_test.go
Comment (achettyiitr) [reply warehouse/warehouse_test.go:165]: Currently, I am just verifying whether it is able to serve HTTP service.

Comment (achettyiitr) [reply warehouse/warehouse_test.go:165]: But, currently, I am not sure how I can verify other things.

File: warehouse/upload.go
Comment (fracasula) [inline warehouse/upload.go:94]: Is this used at all? :thinking:

Comment (achettyiitr) [reply warehouse/upload.go:94]: Good catch. It's not being used.

Comment (achettyiitr) [reply warehouse/upload.go:94]: Curious, do you find it via some static check?

**Comment by achettyiitr** [comment]
> > It still has nothing to do with triggers though so maybe we can give it a generic name or remove it completely.
> > Thoughts?
> 
> Yeah, we can remove this and just pass the dependency for `triggerStore: syncs.Map`. But I don't want to expose **Store, Delete, and Load** to other modules. Currently, it hides the implementation details. Thoughts?

Thanks, @fracasula, for pointing it out regarding `triggerStore`. No need to have it, since it is just proxying to `syncs.Map`

**Comment by achettyiitr** [comment]
Trigger store to use local map.

**Comment by atzoum** [comment]
@achettyiitr this could be a good chance to fix deprecated usages of the config api in warehouse, e.g. `config.RegisterXxxConfigVariable`.

All other packages have been adapted except for `reporting`, `event_schema` (soon to be removed) and `warehouse`

**Comment by achettyiitr** [comment]
> @achettyiitr this could be a good chance to fix deprecated usages of the config api in warehouse, e.g. `config.RegisterXxxConfigVariable`.
> 
> All other packages have been adapted except for `reporting`, `event_schema` (soon to be removed) and `warehouse`

Picking it up just after these changes get merged.

File: warehouse/jobs/runner.go
Comment (fracasula) [inline warehouse/jobs/runner.go:112]:
```suggestion
// Run Async Job runner's main job is to
```

**Review by fracasula — APPROVED** [review]
Approving with the last batch of minor comments. Thanks @achettyiitr for the effort! This is good work :+1:

File: warehouse/warehouse_test.go
Comment (achettyiitr) [reply warehouse/warehouse_test.go:113]: Currently, I have added HTTP and GRPC. We can add incrementally.

**Review by lvrach — APPROVED** [review]

**Pull request merged**
